csharp/src/Google.Protobuf/CodedInputStream.cs

```diff
@@ -584,6 +589,10 @@ public void ReadMessage(IMessage builder)
             }
             int oldLimit = PushLimit(length);
             ++recursionDepth;
+            if (builder is IExtensionMessage)
+            {
+                ExtensionRegistry?.RegisterExtensionsFor(builder as IExtensionMessage);

```

Comment:
Why need to register all the extensions to this message? Can ExtensionSet.ValuesByNumber only contain the extensions that are/were set/get?

Reply:
I believe the reason I did that (it's been nearly a year at this point) was because I needed a way to implement extensions for newly created messages. But now that you mention it, it does have the unintended side effect of adding extensions not added yet which may be good or bad depending on how you look at it.

If this is something we don't want ReadMessage to do, I know a way I can changed it so extensions can only be registered for newly created messages.

Reply:
Some of our internal .proto usages has a lot of extensions but only a few are used. It it a waste to make each new message has such a big dict. 
Have you checked how java implemented it?

Reply:
You're right, I never considered many extension defined for a type where only a few are used. I checked Java's implementation and redesigned ExtensionRegistry and ExtensionSet to more closely match their functionality. Now extensions are only registered from a registry when encountered while reading from the input.